src/lib.rs

```diff
@@ -15,11 +15,98 @@ use scale::{Decode, Encode};
 use std::ops::AddAssign;
 pub use subshard::*;
 
+#[cfg(feature = "parallel")]
+use rayon::prelude::*;
+
+#[cfg(feature = "parallel")]
+use std::sync::Once;
+
+#[cfg(feature = "arena")]
+use bumpalo::Bump;
+
+// Prefetch hints for cache locality optimization
+#[cfg(target_arch = "x86")]
+use std::arch::x86::*;
+#[cfg(target_arch = "x86_64")]
+use std::arch::x86_64::*;
+
+// Branch prediction hints
+#[cold]
+#[inline(never)]
+fn cold() {}
+
+#[inline(always)]
+fn likely(b: bool) -> bool {
+	if !b {
+		cold();
+	}
+	b
+}
+
+#[inline(always)]
+fn unlikely(b: bool) -> bool {
+	if b {
+		cold();
+	}
+	b
+}
+
 pub const MAX_CHUNKS: u16 = 16384;
 
 // The reed-solomon library requires each shards to be 64 bytes aligned.
 const SHARD_ALIGNMENT: usize = 64;
 
+/// Initialize the thread pool for parallel computations.
+/// The number of threads can be configured via the `RAYON_NUM_THREADS` environment variable
+/// at compile time. If the variable is not set, half of the logical CPU cores is used by default.
+/// If `RAYON_NUM_THREADS=0`, all logical CPU cores are used.
+///
+/// Usage example:
+/// ```bash
+/// RAYON_NUM_THREADS=4 cargo build --features parallel
+/// ```
+#[cfg(feature = "parallel")]
+pub(crate) fn init_rayon_thread_pool() {
+	static INIT: Once = Once::new();
+	INIT.call_once(|| {
+		// Helper function to compute default (half of cores)
+		let default_threads = || {
+			let logical_cores = std::thread::available_parallelism().map(|n| n.get()).unwrap_or(1);
+			(logical_cores / 2).max(1)
+		};
+
+		match option_env!("RAYON_NUM_THREADS") {
+			Some(num_threads_str) => {
+				// Variable is set, parse it
+				if let Ok(num_threads) = num_threads_str.parse::<usize>() {
+					if num_threads == 0 {
+						// RAYON_NUM_THREADS=0 means use all CPU cores (rayon default)
+						// Don't set num_threads, let rayon use default
+						let _ = rayon::ThreadPoolBuilder::new().build_global();
+					} else {
+						// RAYON_NUM_THREADS=N (N > 0) means use N threads
+						// Set the environment variable at runtime,
+						// so rayon can use it during initialization
+						std::env::set_var("RAYON_NUM_THREADS", num_threads_str);
+						let _ =
+							rayon::ThreadPoolBuilder::new().num_threads(num_threads).build_global();
```

Comment:
This should not be on the global pool since would affect every other user of rayon, can we use a custom pool here ? 

And this should be somehow modifiable at runtime, because it would be a pain in the ass to have to build and distribute different binaries.